cpp/src/arrow/csv/writer.cc

```diff
@@ -126,13 +126,14 @@ class ColumnPopulator {
 
 // Copies the contents of to out properly escaping any necessary characters.
 // Returns the position prior to last copied character (out_end is decremented).
-char* EscapeReverse(arrow::util::string_view s, char* out_end) {
+char* EscapeReverse(arrow::util::string_view s, char* out_end, bool escaping,
+                    char escape_char) {
   for (const char* val = s.data() + s.length() - 1; val >= s.data(); val--, out_end--) {
-    if (*val == '"') {
-      *out_end = *val;
+    *out_end = *val;
+    if (escaping && *val == '"') {
```

Comment:
Shouldn't we also test for the escape char?

Reply:
Also CountEscapes (or CountQuotes) will also need to test for the escape char.

Reply:
(Though, once merged with #11849: that PR uses CountQuotes to check that quotes don't appear when quoting is disabled, however, we need to make sure that doesn't cause spurious errors if the escape character appears in an unquoted cell. So looks like we need both CountQuotes and CountEscapes.)

Reply:
https://www.rdocumentation.org/packages/readr/versions/1.3.1/topics/write_delim
If I understand correctly, the R function we're trying to align with only escapes quotes.

(Correspond line in the Vroom source: https://github.com/r-lib/vroom/blob/c5b115b1b5f21852886d6407317386b215706239/src/vroom_write.cc#L146)

Reply:
Ah, I see. Then that simplifies things greatly.